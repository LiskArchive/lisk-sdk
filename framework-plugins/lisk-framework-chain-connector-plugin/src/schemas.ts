/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { chain, aggregateCommitSchema, ccmSchema, certificateSchema } from 'lisk-sdk';
import { CCU_FREQUENCY, CCU_TOTAL_CCM_SIZE } from './constants';

const pluginSchemaIDPrefix = '/lisk/plugins/chainConnector';

export const configSchema = {
	$id: `${pluginSchemaIDPrefix}/config`,
	type: 'object',
	properties: {
		receivingChainIPCPath: {
			type: 'string',
			description: 'The IPC path of a receiving node',
		},
		receivingChainWsURL: {
			type: 'string',
			description: 'The WS url of a receiving node',
		},
		ccuFrequency: {
			type: 'integer',
			description: 'Number of blocks after which a CCU should be created',
		},
		encryptedPrivateKey: {
			type: 'string',
			description: 'Encrypted privateKey of the relayer',
		},
		ccuFee: {
			type: 'string',
			format: 'uint64',
			description: 'Fee to be paid for each CCU transaction',
		},
		password: {
			type: 'string',
			description: 'Password to decrypt encryptedPrivateKey',
		},
		saveCCU: {
			type: 'boolean',
			description:
				'Flag for the user to either save or send a CCU on creation. Send is by default.',
		},
		maxCCUSize: {
			type: 'integer',
			description: 'Maximum size of CCU to be allowed',
			minimum: 1,
			maximum: CCU_TOTAL_CCM_SIZE,
		},
	},
	required: ['ccuFee', 'encryptedPrivateKey', 'password'],
	default: {
		ccuFrequency: CCU_FREQUENCY,
		isSaveCCU: false,
		maxCCUSize: CCU_TOTAL_CCM_SIZE,
	},
};

export const validatorsDataSchema = {
	$id: `${pluginSchemaIDPrefix}/validatorsData`,
	type: 'object',
	required: ['validators', 'certificateThreshold'],
	properties: {
		validators: {
			type: 'array',
			fieldNumber: 1,
			items: {
				type: 'object',
				required: ['blsKey', 'bftWeight'],
				properties: {
					address: { dataType: 'bytes', fieldNumber: 1 },
					blsKey: { dataType: 'bytes', fieldNumber: 2 },
					bftWeight: { dataType: 'uint64', fieldNumber: 3 },
				},
			},
		},
		certificateThreshold: { dataType: 'uint64', fieldNumber: 2 },
		validatorsHash: { dataType: 'bytes', fieldNumber: 3 },
	},
};

export const blockHeadersInfoSchema = {
	$id: `${pluginSchemaIDPrefix}/blockHeaders`,
	type: 'object',
	properties: {
		blockHeaders: {
			type: 'array',
			fieldNumber: 1,
			items: {
				...chain.blockHeaderSchema,
			},
		},
	},
};

export const aggregateCommitsInfoSchema = {
	$id: `${pluginSchemaIDPrefix}/aggregateCommits`,
	type: 'object',
	properties: {
		aggregateCommits: {
			type: 'array',
			fieldNumber: 1,
			items: {
				...aggregateCommitSchema,
			},
		},
	},
};

export const validatorsHashPreimageInfoSchema = {
	$id: `${pluginSchemaIDPrefix}/validatorsHashPreimage`,
	type: 'object',
	properties: {
		validatorsHashPreimage: {
			type: 'array',
			fieldNumber: 1,
			items: {
				...validatorsDataSchema,
			},
		},
	},
};

export const lastSentCCMWithHeight = {
	$id: `${pluginSchemaIDPrefix}/lastSentCCMWithHeight`,
	type: 'object',
	properties: {
		...ccmSchema.properties,
		height: { dataType: 'uint32', fieldNumber: Object.keys(ccmSchema.properties).length + 1 },
	},
};

export const listOfCCUsSchema = {
	$id: `${pluginSchemaIDPrefix}/listOfCCUs`,
	type: 'object',
	properties: {
		listOfCCUs: {
			type: 'array',
			fieldNumber: 1,
			items: {
				...chain.transactionSchema,
			},
		},
	},
};

export const ccmsFromEventsSchema = {
	$id: `${pluginSchemaIDPrefix}/ccmsFromEvents`,
	type: 'object',
	properties: {
		ccmsFromEvents: {
			type: 'array',
			fieldNumber: 1,
			items: {
				type: 'object',
				properties: {
					ccms: {
						type: 'array',
						fieldNumber: 1,
						items: {
							...ccmSchema,
						},
					},
					height: { dataType: 'uint32', fieldNumber: 2 },
					inclusionProof: {
						type: 'object',
						fieldNumber: 3,
						properties: {
							siblingHashes: {
								type: 'array',
								fieldNumber: 1,
								items: {
									dataType: 'bytes',
								},
							},
							bitmap: {
								dataType: 'bytes',
								fieldNumber: 2,
							},
						},
					},
				},
			},
		},
	},
};
