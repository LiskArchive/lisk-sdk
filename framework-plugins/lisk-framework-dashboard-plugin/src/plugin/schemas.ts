/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const configSchema = {
	$id: '#/plugins/liskDashboard/config',
	type: 'object',
	properties: {
		applicationName: {
			type: 'string',
			description: 'Application name to be shown near Logo',
		},
		applicationUrl: {
			type: 'string',
			format: 'uri',
			description: 'URL to connect',
		},
		port: {
			type: 'integer',
			minimum: 1,
			maximum: 65535,
		},
		host: {
			type: 'string',
			format: 'ip',
		},
	},
	required: [],
	default: {
		applicationUrl: 'ws://127.0.0.1:7887/rpc-ws',
		port: 4005,
		host: '127.0.0.1',
		applicationName: 'Lisk',
	},
};
