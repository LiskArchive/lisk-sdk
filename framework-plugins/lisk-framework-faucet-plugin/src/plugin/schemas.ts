/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const configSchema = {
	$id: '#/plugins/liskFaucet/config',
	type: 'object',
	properties: {
		port: {
			type: 'integer',
			minimum: 1,
			maximum: 65535,
		},
		host: {
			type: 'string',
			format: 'ip',
		},
		encryptedPrivateKey: {
			type: 'string',
			description: 'Encrypted private key of the faucet account',
		},
		tokenID: {
			type: 'string',
			format: 'hex',
			description: 'TokenID of faucet',
		},
		applicationUrl: {
			type: 'string',
			format: 'uri',
			description: 'URL to connect',
		},
		fee: {
			type: 'string',
			description: 'The transaction fee used to faucet an account',
		},
		amount: {
			type: 'string',
			description: 'Number of tokens to fund an account per request',
		},
		tokenPrefix: {
			type: 'string',
			description: 'The token prefix associated with your application',
		},
		logoURL: {
			type: 'string',
			format: 'uri',
			description: 'The URL of the logo used on the UI',
		},
		captchaSecretkey: {
			type: 'string',
			description: 'The re-captcha secret key',
		},
		captchaSitekey: {
			type: 'string',
			description: 'The re-captcha site key',
		},
	},
	required: ['tokenID', 'encryptedPrivateKey', 'captchaSecretkey', 'captchaSitekey'],
	default: {
		port: 4004,
		host: '127.0.0.1',
		applicationUrl: 'ws://localhost:7887/rpc-ws',
		fee: '0.1',
		amount: '100',
		tokenPrefix: 'lsk',
	},
};

export const fundParamsSchema = {
	$id: '/lisk/faucet/fund',
	type: 'object',
	required: ['address'],
	properties: {
		address: {
			type: 'string',
			format: 'hex',
		},
		token: {
			type: 'string',
		},
	},
};

export const authorizeParamsSchema = {
	$id: '/lisk/faucet/auth',
	type: 'object',
	required: ['password', 'enable'],
	properties: {
		password: {
			type: 'string',
		},
		enable: {
			type: 'boolean',
		},
	},
};
