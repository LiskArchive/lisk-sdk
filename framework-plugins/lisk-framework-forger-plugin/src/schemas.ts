/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const forgerInfoSchema = {
	$id: '/forger/info',
	type: 'object',
	properties: {
		totalProducedBlocks: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		totalReceivedFees: {
			dataType: 'uint64',
			fieldNumber: 2,
		},
		totalReceivedRewards: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
		stakeReceived: {
			type: 'array',
			fieldNumber: 4,
			items: {
				type: 'object',
				properties: {
					address: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					amount: {
						dataType: 'uint64',
						fieldNumber: 2,
					},
				},
			},
			required: ['address', 'amount'],
		},
	},
	required: ['totalProducedBlocks', 'totalReceivedFees', 'totalReceivedRewards', 'stakeReceived'],
};

export const forgerSyncSchema = {
	$id: '/forger/sync',
	type: 'object',
	required: ['syncUptoHeight'],
	properties: {
		syncUptoHeight: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
	},
};

export const posVoteParamsSchema = {
	$id: '/forger/pos/stake',
	type: 'object',
	required: ['stakes'],
	properties: {
		stakes: {
			type: 'array',
			minItems: 1,
			maxItems: 20,
			items: {
				type: 'object',
				required: ['validatorAddress', 'amount'],
				properties: {
					validatorAddress: {
						dataType: 'bytes',
						fieldNumber: 1,
						minLength: 20,
						maxLength: 20,
					},
					amount: {
						dataType: 'sint64',
						fieldNumber: 2,
					},
				},
			},
			fieldNumber: 1,
		},
	},
};
