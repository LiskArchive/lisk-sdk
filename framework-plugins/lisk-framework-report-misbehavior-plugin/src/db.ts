/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import * as os from 'os';
import { join } from 'path';
import { ensureDir } from 'fs-extra';
import { cryptography, codec, chain, db as liskDB, BasePlugin } from 'lisk-sdk';

const { BlockHeader } = chain;
const { utils } = cryptography;

export const blockHeadersSchema = {
	$id: '/lisk/reportMisbehavior/blockHeaders',
	type: 'object',
	required: ['blockHeaders'],
	properties: {
		blockHeaders: {
			type: 'array',
			fieldNumber: 1,
			items: {
				dataType: 'bytes',
			},
		},
	},
};

export interface BlockHeaderAsset {
	readonly seedReveal: Buffer;
	readonly maxHeightPreviouslyForged: number;
	readonly maxHeightPrevoted: number;
}

interface BlockHeaders {
	readonly blockHeaders: Buffer[];
}

export const getDBInstance = async (
	dataPath: string,
	dbName = 'lisk-framework-report-misbehavior-plugin.db',
): Promise<liskDB.Database> => {
	const dirPath = join(dataPath.replace('~', os.homedir()), 'plugins/data', dbName);
	await ensureDir(dirPath);

	return new liskDB.Database(dirPath);
};

export const getBlockHeaders = async (
	db: liskDB.Database,
	dbKeyBlockHeader: Buffer,
): Promise<BlockHeaders> => {
	try {
		const encodedBlockHeaders = await db.get(dbKeyBlockHeader);
		return codec.decode<BlockHeaders>(blockHeadersSchema, encodedBlockHeaders);
	} catch (error) {
		return { blockHeaders: [] as Buffer[] };
	}
};

export const saveBlockHeaders = async (
	db: liskDB.Database,
	headerBytes: Buffer,
): Promise<boolean> => {
	const header = BlockHeader.fromBytes(headerBytes);
	const heightBytes = Buffer.alloc(4);
	heightBytes.writeUInt32BE(header.height, 0);
	const dbKey = Buffer.concat([header.generatorAddress, Buffer.from(':', 'utf8'), heightBytes]);
	const { blockHeaders } = await getBlockHeaders(db, dbKey);

	if (!blockHeaders.find(blockHeader => utils.hash(blockHeader).equals(header.id))) {
		await db.set(
			dbKey,
			codec.encode(blockHeadersSchema, {
				blockHeaders: [...blockHeaders, header.getBytes()],
			}),
		);
		return true;
	}
	return false;
};

type IteratableStream = NodeJS.ReadableStream & { destroy: (err?: Error) => void };

export const getContradictingBlockHeader = async (
	db: liskDB.Database,
	blockHeader: chain.BlockHeader,
	apiClient: BasePlugin['apiClient'],
): Promise<chain.BlockHeader | undefined> => {
	const header1 = blockHeader.getBytes().toString('hex');
	const existingHeaders = await new Promise<string[]>((resolve, reject) => {
		const stream = db.createReadStream({
			gte: Buffer.concat([blockHeader.generatorAddress, Buffer.alloc(4, 0)]),
			lte: Buffer.concat([blockHeader.generatorAddress, Buffer.alloc(4, 255)]),
		}) as IteratableStream;
		const results: string[] = [];
		stream
			.on('data', ({ value }: { value: Buffer }) => {
				const { blockHeaders } = codec.decode<BlockHeaders>(blockHeadersSchema, value);
				for (const encodedHeader of blockHeaders) {
					results.push(encodedHeader.toString('hex'));
				}
			})
			.on('error', error => {
				reject(error);
			})
			.on('end', () => {
				resolve(results);
			});
	});
	for (const header2 of existingHeaders) {
		const contradicting = await apiClient.invoke('bft_areHeadersContradicting', {
			header1,
			header2,
		});
		if (contradicting) {
			return BlockHeader.fromBytes(Buffer.from(header2, 'hex'));
		}
	}
	return undefined;
};

export const clearBlockHeaders = async (
	db: liskDB.Database,
	currentHeight: number,
): Promise<void> => {
	const keys = await new Promise<Buffer[]>((resolve, reject) => {
		const stream = db.createReadStream();
		const res: Buffer[] = [];
		stream
			.on('data', ({ key, value }: { key: Buffer; value: Buffer }) => {
				const { blockHeaders } = codec.decode<BlockHeaders>(blockHeadersSchema, value);
				for (const encodedHeader of blockHeaders) {
					const decodedBlockHeader = BlockHeader.fromBytes(encodedHeader);
					if (decodedBlockHeader.height < currentHeight - 260000) {
						res.push(key);
					}
				}
			})
			.on('error', error => {
				reject(error);
			})
			.on('end', () => {
				resolve(res);
			});
	});
	const batch = new liskDB.Batch();
	for (const k of keys) {
		batch.del(k);
	}
	await db.write(batch);
};
