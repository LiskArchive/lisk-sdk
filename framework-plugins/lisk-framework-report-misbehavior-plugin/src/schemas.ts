/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const configSchema = {
	$id: '#/plugins/lisk-report-misbehavior/config',
	type: 'object',
	properties: {
		clearBlockHeadersInterval: {
			type: 'integer',
			description: 'Frequency of cleaning up the outdated data in second.',
		},
		encryptedPassphrase: {
			type: 'string',
			format: 'encryptedPassphrase',
		},
		defaultPassword: {
			type: 'string',
		},
		fee: {
			type: 'integer',
			description: 'The fee required to report misbehavior transaction.',
		},
	},
	required: [],
	default: {
		clearBlockHeadersInterval: 60000,
		encryptedPassphrase: '',
		defaultPassword: '',
		fee: 100000000,
	},
};

export const actionParamsSchema = {
	$id: 'lisk/report_misbehavior/auth',
	type: 'object',
	required: ['password', 'enable'],
	properties: {
		password: {
			type: 'string',
		},
		enable: {
			type: 'boolean',
		},
	},
};
