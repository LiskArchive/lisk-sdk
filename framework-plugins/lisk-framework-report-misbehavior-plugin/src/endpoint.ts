/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import {
	BasePluginEndpoint,
	PluginEndpointContext,
	validator as liskValidator,
	cryptography,
} from 'lisk-sdk';
import { actionParamsSchema } from './schemas';
import { ReportMisbehaviorPluginConfig, State } from './types';

const { validator} = liskValidator;
const { encrypt, address } = cryptography;

export class Endpoint extends BasePluginEndpoint {
	private _state!: State;
	private _config!: ReportMisbehaviorPluginConfig;

	public init(state: State, config: ReportMisbehaviorPluginConfig) {
		this._state = state;
		this._config = config;
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async authorize(context: PluginEndpointContext): Promise<{ result: string }> {
		validator.validate(actionParamsSchema, context.params);

		const { enable, password } = context.params;

		try {
			const parsedEncryptedPassphrase = encrypt.parseEncryptedPassphrase(
				this._config.encryptedPassphrase,
			);

			const passphrase = await encrypt.decryptPassphraseWithPassword(
				parsedEncryptedPassphrase as any,
				password as string,
			);

			const { publicKey } = address.getAddressAndPublicKeyFromPassphrase(passphrase);

			this._state.publicKey = enable ? publicKey : undefined;
			this._state.passphrase = enable ? passphrase : undefined;
			const changedState = enable ? 'enabled' : 'disabled';

			return {
				result: `Successfully ${changedState} the reporting of misbehavior.`,
			};
		} catch (error) {
			throw new Error('Password given is not valid.');
		}
	}
}
