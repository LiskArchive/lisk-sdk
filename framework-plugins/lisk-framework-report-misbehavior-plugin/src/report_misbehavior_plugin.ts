/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import {
	BasePlugin,
	PluginInitContext,
	db as liskDB,
	codec,
	chain,
	cryptography,
	blockHeaderSchema,
} from 'lisk-sdk';
import {
	getDBInstance,
	saveBlockHeaders,
	getContradictingBlockHeader,
	clearBlockHeaders,
} from './db';
import { ReportMisbehaviorPluginConfig, State } from './types';
import { configSchema } from './schemas';
import { Endpoint } from './endpoint';

const { getAddressAndPublicKeyFromPassphrase, getAddressFromPassphrase, signData } = cryptography;
const { BlockHeader, Transaction, TAG_TRANSACTION } = chain;

export class ReportMisbehaviorPlugin extends BasePlugin<ReportMisbehaviorPluginConfig> {
	public name = 'reportMisbehavior';
	public configSchema = configSchema;
	public endpoint = new Endpoint();

	private _pluginDB!: liskDB.KVStore;
	private readonly _state: State = { currentHeight: 0 };
	private _clearBlockHeadersIntervalId!: NodeJS.Timer | undefined;

	public get nodeModulePath(): string {
		return __filename;
	}

	public async init(context: PluginInitContext): Promise<void> {
		await super.init(context);
		this.endpoint.init(this._state, this.config);
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async load(): Promise<void> {
		// TODO: https://github.com/LiskHQ/lisk-sdk/issues/6201
		this._pluginDB = await getDBInstance(this.dataPath);
		// Listen to new block and delete block events
		this._subscribeToChannel();
		// eslint-disable-next-line @typescript-eslint/no-misused-promises
		this._clearBlockHeadersIntervalId = setInterval(() => {
			clearBlockHeaders(this._pluginDB, this._state.currentHeight).catch(error =>
				this.logger.error(error),
			);
		}, this.config.clearBlockHeadersInterval);
	}

	public async unload(): Promise<void> {
		clearInterval(this._clearBlockHeadersIntervalId as NodeJS.Timer);

		await this._pluginDB.close();
	}

	private _subscribeToChannel(): void {
		this.apiClient.subscribe(
			'network_newBlock',
			async (event: Record<string, unknown> | undefined) => {
				if (!event) {
					this.logger.error('Invalid payload for network_newBlock');
					return;
				}

				const { blockHeader: blockHeaderJSON } = event;

				const headerBytes = codec.encodeJSON(
					blockHeaderSchema,
					blockHeaderJSON as Record<string, unknown>,
				);
				try {
					const saved = await saveBlockHeaders(this._pluginDB, headerBytes);
					if (!saved) {
						return;
					}
					const decodedBlockHeader = BlockHeader.fromBytes(headerBytes);

					// Set new currentHeight
					if (decodedBlockHeader.height > this._state.currentHeight) {
						this._state.currentHeight = decodedBlockHeader.height;
					}
					const contradictingBlock = await getContradictingBlockHeader(
						this._pluginDB,
						decodedBlockHeader,
						this.apiClient,
					);
					if (contradictingBlock && this._state.passphrase) {
						const encodedTransaction = await this._createPoMTransaction(
							decodedBlockHeader,
							contradictingBlock,
						);
						const result = await this.apiClient.invoke<{
							transactionId?: string;
						}>('app_postTransaction', {
							transaction: encodedTransaction,
						});

						this.logger.debug('Sent Report misbehavior transaction', result.transactionId);
					}
				} catch (error) {
					this.logger.error(error);
				}
			},
		);
	}

	private async _createPoMTransaction(
		contradictingBlock: chain.BlockHeader,
		decodedBlockHeader: chain.BlockHeader,
	): Promise<string> {
		// ModuleID:13 (DPoS), CommandID:3 (PoMCommand)
		const dposMeta = this.apiClient.metadata.find(m => m.id === 13);
		if (!dposMeta) {
			throw new Error('DPoS module is not registered in the application.');
		}
		const pomParamsInfo = dposMeta.commands.find(m => m.id === 3);
		if (!pomParamsInfo || !pomParamsInfo.params) {
			throw new Error('PoM params schema is not registered in the application.');
		}

		// Assume passphrase is checked before calling this function
		const passphrase = this._state.passphrase as string;

		const authAccount = await this.apiClient.invoke<{ nonce: string }>('auth_getAuthAccount', {
			address: getAddressFromPassphrase(passphrase).toString('hex'),
		});

		const pomTransactionParams = {
			header1: decodedBlockHeader,
			header2: contradictingBlock,
		};

		const { networkIdentifier } = await this.apiClient.invoke<{ networkIdentifier: string }>(
			'system_getNodeInfo',
		);

		const encodedParams = codec.encode(pomParamsInfo.params, pomTransactionParams);

		const tx = new Transaction({
			moduleID: dposMeta.id,
			commandID: pomParamsInfo.id,
			nonce: BigInt(authAccount.nonce),
			senderPublicKey:
				this._state.publicKey ?? getAddressAndPublicKeyFromPassphrase(passphrase).publicKey,
			fee: BigInt(this.config.fee), // TODO: The static fee should be replaced by fee estimation calculation
			params: encodedParams,
			signatures: [],
		});

		tx.signatures.push(
			signData(
				TAG_TRANSACTION,
				Buffer.from(networkIdentifier, 'hex'),
				tx.getSigningBytes(),
				passphrase,
			),
		);

		return tx.getBytes().toString('hex');
	}
}
