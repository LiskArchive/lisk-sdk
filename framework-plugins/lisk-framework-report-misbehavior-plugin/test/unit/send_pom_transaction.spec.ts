/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { utils } from '@liskhq/lisk-cryptography';
import { testing, chain, ApplicationConfigForPlugin } from 'lisk-sdk';
import { when } from 'jest-when';

import { ReportMisbehaviorPlugin } from '../../src';
import { configSchema } from '../../src/schemas';

const appConfigForPlugin: ApplicationConfigForPlugin = {
	...testing.fixtures.defaultConfig,
};

const validPluginOptions = {
	...configSchema.default,
	encryptedPassphrase:
		'salt=683425ca06c9ff88a5ab292bb5066dc5&cipherText=4ce151&iv=bfaeef79a466e370e210f3c6&tag=e84bf097b1ec5ae428dd7ed3b4cce522&version=1',
	dataPath: '/my/app',
};

describe('Send PoM transaction', () => {
	let reportMisbehaviorPlugin: ReportMisbehaviorPlugin;
	const defaultchainID = '93d00fe5be70d90e7ae247936a2e7d83b50809c79b73fa14285f02c842348b3e';
	const random32Bytes = Buffer.from(
		'3d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e65898990',
		'hex',
	);
	const random20Bytes = Buffer.from('40ff452fae2affe6eeef3c30e53e9eac35a1bc43', 'hex');
	const channelMock = {
		registerToBus: jest.fn(),
		once: jest.fn(),
		publish: jest.fn(),
		subscribe: jest.fn(),
		isValidEventName: jest.fn(),
		isValidActionName: jest.fn(),
		invoke: jest.fn(),
		eventsList: [],
		actionsList: [],
		actions: {},
		moduleName: '',
		options: {},
	} as any;
	const header1 = new chain.BlockHeader({
		height: 100,
		aggregateCommit: {
			aggregationBits: Buffer.alloc(0),
			certificateSignature: Buffer.alloc(0),
			height: 0,
		},
		impliesMaxPrevotes: false,
		generatorAddress: random20Bytes,
		maxHeightGenerated: 0,
		maxHeightPrevoted: 50,
		previousBlockID: random32Bytes,
		timestamp: 100,
		version: 2,
		assetRoot: random32Bytes,
		eventRoot: random32Bytes,
		stateRoot: random32Bytes,
		transactionRoot: random32Bytes,
		validatorsHash: random32Bytes,
		signature: random32Bytes,
	});
	const header2 = new chain.BlockHeader({
		...header1.toObject(),
		height: 100,
		maxHeightPrevoted: 32,
	});

	beforeEach(async () => {
		reportMisbehaviorPlugin = new ReportMisbehaviorPlugin();
		reportMisbehaviorPlugin['_apiClient'] = {
			schema: {
				block: chain.blockSchema,
				header: chain.blockHeaderSchema,
				transaction: chain.transactionSchema,
			},
			metadata: [
				{
					id: utils.intToBuffer(13, 4).toString('hex'),
					name: 'pos',
					commands: [
						{
							id: utils.intToBuffer(3, 4).toString('hex'),
							name: 'reportDelegateMisbehavior',
							params: {
								$id: '/lisk/pos/pom',
								type: 'object',
								required: ['header1', 'header2'],
								properties: {
									header1: {
										...chain.blockHeaderSchema,
										$id: 'blockHeader1',
										fieldNumber: 1,
									},
									header2: {
										...chain.blockHeaderSchema,
										$id: 'blockHeader2',
										fieldNumber: 2,
									},
								},
							},
						},
					],
				},
			],
			invoke: jest.fn(),
			subscribe: jest.fn(),
		};
		await reportMisbehaviorPlugin.init({
			config: validPluginOptions,
			appConfig: appConfigForPlugin,
			logger: testing.mocks.loggerMock,
		});
		(reportMisbehaviorPlugin as any)._channel = channelMock;
		(reportMisbehaviorPlugin as any)._options = { fee: '100000000' };
		reportMisbehaviorPlugin['logger'] = {
			error: jest.fn(),
		} as any;
		(reportMisbehaviorPlugin as any)._state = {
			privateKey: Buffer.from(testing.fixtures.defaultFaucetAccount.privateKey, 'hex'),
			publicKey: Buffer.from(testing.fixtures.defaultFaucetAccount.publicKey, 'hex'),
		};

		when(jest.spyOn(reportMisbehaviorPlugin['apiClient'], 'invoke'))
			.calledWith('auth_getAuthAccount', expect.anything())
			.mockResolvedValue({ nonce: '0' } as never)
			.calledWith('system_getNodeInfo')
			.mockResolvedValue({ chainID: defaultchainID } as never);
	});

	it('should throw error when pom transaction params schema is not found', async () => {
		reportMisbehaviorPlugin['_apiClient'] = {
			schema: {
				block: chain.blockSchema,
				header: chain.blockHeaderSchema,
				transaction: chain.transactionSchema,
			},
			metadata: [
				{
					id: utils.intToBuffer(13, 4).toString('hex'),
					name: 'pos',
					commands: [],
				},
			],
			invoke: jest.fn(),
			subscribe: jest.fn(),
		};
		await expect(
			(reportMisbehaviorPlugin as any)._createPoMTransaction(header1, header2),
		).rejects.toThrow('PoM params schema is not registered in the application.');
	});

	it('should create pom transaction for given block headers', async () => {
		const pomTransaction = await (reportMisbehaviorPlugin as any)._createPoMTransaction(
			header1,
			header2,
		);
		expect(pomTransaction).toMatchSnapshot();
	});
});
