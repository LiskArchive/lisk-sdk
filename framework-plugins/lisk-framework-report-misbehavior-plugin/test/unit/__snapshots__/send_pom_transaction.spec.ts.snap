// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send PoM transaction should create pom transaction for given block headers 1`] = `"080d100318002080c2d72f2a200fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a32b2040a960208021064186422203d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e658989902a1440ff452fae2affe6eeef3c30e53e9eac35a1bc4332203d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e658989903a203d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e6589899042203d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e658989904a203d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e658989905020580062203d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e658989906a06080012001a0072203d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e6589899012960208021064186422203d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e658989902a1440ff452fae2affe6eeef3c30e53e9eac35a1bc4332203d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e658989903a203d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e6589899042203d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e658989904a203d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e658989905032580062203d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e658989906a06080012001a0072203d1b5dd1ef4ff7b22359598ebdf58966a51adcc03e02ad356632743e658989903a405c62a402ef7cf82d0c67a0d9fd4ecb31334ddb37509c6aed8e7e7bbe4918e52c387f1f9ae6e66bbdb53f41e785a7c0034aee74578403a1d580828855b6be970e"`;
